### PR TITLE
Nlp candidate squad

### PR DESCRIPTION
**Pull request by yzhang123**
* added training & evaluation of squad (so far just loss)
* added ci tests

### CODE REVIEW

File: nemo/collections/nlp/modules/common/token_classifier.py [hunk 1]
```diff
@@ -70,6 +70,88 @@ def __init__(
...
         """
         super().__init__()
+        self.mlp = MultiLayerPerceptron(
+            hidden_size, num_classes, num_layers=num_layers, activation=activation, log_softmax=log_softmax
+        )
+        self.dropout = nn.Dropout(dropout)
+        if use_transformer_init:
+            self.apply(lambda module: transformer_weights_init(module, xavier=False))
+
+    @typecheck()
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/token_classifier.py:80]: could you please check default values for TokenClassifier? num_layers is 2 as far as I remember, smth else could be also different, these defaults are from BERTTokenClassifier

File: nemo/collections/nlp/modules/common/token_classifier.py [hunk 2]
```diff
@@ -70,6 +70,88 @@ def __init__(
...
+        """
+        pass
+
+    @classmethod
+    def restore_from(cls, restore_path: str):
+        """
+        Restores the module from the specified path.
+        Args:
+            :param restore_path: Path to restore the module from.
+        """
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/token_classifier.py:107]: :param and :return I could be the who added them actually, would appreciate if you delete them

File: Jenkinsfile [hunk 3]
```diff
@@ -33,6 +33,48 @@ pipeline {
...
+    stage('L2: BERT Squad v1.1') {
+      when {
+        anyOf{
+          branch 'candidate'
+          changeRequest()
+        }
+      }
+      failFast true
+        steps {
+          sh 'cd examples/nlp && CUDA_VISIBLE_DEVICES=0 python question_answering_squad.py --precision 16 --amp_level=O1 --train_file /home/TestData/nlp/squad_mini/v1.1/train-v1.1.json --eval_file /home/TestData/nlp/squad_mini/v1.1/dev-v1.1.json --batch_size 8 --gpus 1 --do_lower_case --pretrained_model_name bert-base-uncased --optimizer adamw --lr 5e-5 --max_steps 50 --scheduler WarmupAnnealing'
```
Comment (okuchaiev) [inline Jenkinsfile:45]: Thanks for adding tests! 
Do we really need 50 and 10 steps? Can we do say 5 and/or just 2 to have reasonable confidence that nothing crashes?

File: nemo/collections/common/losses/spanning_loss.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            "end_positions": NeuralType(tuple('B'), ChannelType()),
+        }
+
+    @property
+    def output_ports(self):
+        """Returns definitions of module output ports.
+        """
+        return {
+            "loss": NeuralType(elements_type=LossType()),
+            "start_logits": NeuralType(('B', 'T'), ChannelType()),
```
Comment (okuchaiev) [inline nemo/collections/common/losses/spanning_loss.py:49]: should start_logits and end_logits be of LogitsType?

File: nemo/collections/nlp/models/qa_model.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        num_layers: int = 1,
+        activation: str = 'relu',
+        log_softmax: bool = False,
+        dropout: float = 0.0,
+        use_transformer_init: bool = True,
+    ):
+        """
+        Args:
+            :param num_classes: number of classes
+            :param pretrained_model_name: pretrained language model name, to see the complete list use
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply nemo/collections/nlp/models/qa_model.py:65]: The format of the docstring is wrong - should be different style (I think called Google Style in Pycharm) - please have a look at other classes (e.g. in the core)

File: nemo/collections/nlp/models/qa_model.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        """
+        # forward pass
+        input_ids, input_type_ids, input_mask, unique_ids, start_positions, end_positions = batch
+        logits = self.forward(input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask)
+        loss, start_logits, end_logits = self.loss(
+            logits=logits, start_positions=start_positions, end_positions=end_positions
+        )
+
+        tensorboard_logs = {'val_loss': loss}
+        return {'val_loss': loss, 'log': tensorboard_logs}
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/qa_model.py:138]: please remove tensorboard_logs from validation_step function. 
log validation loss (and actually other metrics too) after aggregation inside validation_epoch_end function. This way validation loss/metrics are per validation epoch

File: nemo/collections/common/parts/utils.py [hunk 7]
```diff
@@ -31,3 +31,26 @@ def if_exist(outfold: str, files: List[str]):
         if not os.path.exists(f'{outfold}/{file}'):
             return False
     return True
+
+
+def _compute_softmax(scores):
+    """Compute softmax probability over raw logits."""
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/common/parts/utils.py:37]: scores are arrays not tensors

File: examples/nlp/question_answering_squad.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        "--use_cache", action='store_true', help="When specified do not load and store cache preprocessed data.",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    parser = ArgumentParser()
+    # parser = pl.Trainer.add_argparse_args(parser)
+    parser = add_optimizer_args(parser)
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/question_answering_squad.py:149]: no

File: nemo/collections/nlp/modules/common/token_classifier.py [hunk 9]
```diff
@@ -70,6 +70,88 @@ def __init__(
...
+        dropout: float = 0.0,
+        use_transformer_init: bool = True,
+    ) -> None:
+
+        """
+        Initializes the Token Classifier module.
+
+        Args:
+            :param hidden_size: the size of the hidden dimension
+            :param num_classes: number of classes
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/token_classifier.py:147]: done

File: nemo/collections/common/losses/spanning_loss.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+from nemo.core.neural_types import ChannelType, LogitsType, LossType, NeuralType
+from nemo.utils.decorators import experimental
+
+__all__ = ['SpanningLoss']
+
+
+@experimental
+class SpanningLoss(Loss):
+    """
+    implements QuestionAnswering loss.
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/common/losses/spanning_loss.py:30]: done

File: nemo/collections/nlp/data/qa_dataset.py [hunk 11]
```diff
@@ -0,0 +1,805 @@
...
+        doc_stride (int): When splitting up a long document into chunks,
+            how much stride to take between chunks.
+        max_query_length (iny): All training files which have a duration less
+            than min_duration are dropped. Can't be used if the `utt2dur` file
+            does not exist. Defaults to None.
+        max_seq_length (int): All training files which have a duration more
+            than max_duration are dropped. Can't be used if the `utt2dur` file
+            does not exist. Defaults to None.
+        mode (str): Use "train", "eval" or "test" to define between
+            training and evaluation.
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/data/qa_dataset.py:73]: done

File: nemo/collections/nlp/models/qa_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/models/qa_model.py:65]: done

File: nemo/collections/nlp/modules/common/token_classifier.py (hunk 1 above)
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/token_classifier.py:80]: done

File: nemo/collections/nlp/modules/common/token_classifier.py (hunk 2 above)
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/token_classifier.py:107]: done

File: Jenkinsfile (hunk 3 above)
Comment (yzhang123) [reply Jenkinsfile:45]: ok

File: nemo/collections/common/losses/spanning_loss.py (hunk 4 above)
Comment (yzhang123) [reply nemo/collections/common/losses/spanning_loss.py:49]: yes thx

File: nemo/collections/nlp/models/qa_model.py (hunk 6 above)
Comment (yzhang123) [reply nemo/collections/nlp/models/qa_model.py]: done

File: nemo/collections/nlp/modules/common/token_classifier.py [hunk 12]
```diff
@@ -50,7 +50,89 @@ def __init__(
...
+        """
+        Returns definitions of module output ports.
+        """
+        return {"logits": NeuralType(('B', 'T', 'C'), LogitsType())}
+
+    def __init__(
+        self,
+        hidden_size: int,
+        num_classes: int,
+        num_layers: int = 2,
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/token_classifier.py:135]: num_layers was 1 for BERTtokenclassifier

File: nemo/collections/common/parts/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ekmb) [reply nemo/collections/common/parts/utils.py:37]: should it be in common?

File: nemo/collections/common/parts/__init__.py [hunk 13]
```diff
@@ -12,6 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from nemo.collections.common.parts.data_utils import *
 from nemo.collections.common.parts.multi_layer_perceptron import MultiLayerPerceptron
 from nemo.collections.common.parts.transformer_utils import *
+from nemo.collections.common.parts.utils import *
```
Comment (ekmb) [inline nemo/collections/common/parts/__init__.py:17]: could you please remove *

Comment (yzhang123) [reply nemo/collections/common/parts/__init__.py:17]: why? those are utils functions not classes and will grow with time, enumerating all is error prone

File: nemo/collections/nlp/modules/common/token_classifier.py (hunk 12 above)
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/token_classifier.py]: changed

File: nemo/collections/common/parts/__init__.py (hunk 13 above)
Comment (ekmb) [reply nemo/collections/common/parts/__init__.py:17]: I just noticed @blisc started pushing for this in asr or are we fine with * in __init__.py ? @okuchaiev ?

Comment (ekmb) [reply nemo/collections/common/parts/__init__.py:17]: let's merge now, since it's a blocker for other PR, and update later if needed
